### PR TITLE
PERF: pd.concat EA-backed indexes and sort=True

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.0.0.rst` file if fixing a bug or adding a new feature.

`algos.safe_sort` is currently converting EA-backed indexes to ndarrays which can cause a perf hit. It can be significant if the index contains pd.NA as `np.argsort` will raise which gets caught in the try-catch. This PR avoids the numpy conversion for EA-backed indexes. 

Note: One test which relied on the numpy conversion was updated.

```
import numpy as np
import pandas as pd
from pandas.core.indexes.api import safe_sort_index

vals = [pd.NA] + list(np.arange(100_000))
idx = pd.Index(vals, dtype="Int64")

%timeit safe_sort_index(idx)

# 81.6 ms ± 5.29 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)   <- main
# 2.73 ms ± 24.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)  <- PR
```

I updated an existing ASV that was just added yesterday to cover this case as well:

```
       before           after         ratio
     [8b503a8c]       [2e2a02fa]
                      <safe-sort-index>
-        27.1±1ms       22.2±0.4ms     0.82  join_merge.ConcatIndexDtype.time_concat_series('string[pyarrow]', 'non_monotonic', 1, True)
-      15.9±0.2ms       12.3±0.1ms     0.78  join_merge.ConcatIndexDtype.time_concat_series('string[python]', 'has_na', 1, True)
-      31.4±0.7ms       24.1±0.7ms     0.77  join_merge.ConcatIndexDtype.time_concat_series('string[pyarrow]', 'has_na', 1, True)
-      22.8±0.2ms       16.4±0.3ms     0.72  join_merge.ConcatIndexDtype.time_concat_series('Int64', 'has_na', 1, True)
-      13.8±0.4ms       9.26±0.1ms     0.67  join_merge.ConcatIndexDtype.time_concat_series('Int64', 'non_monotonic', 1, True)
```

### CODE REVIEW

File: pandas/core/algorithms.py [hunk 1]
```diff
@@ -1894,15 +1893,18 @@ def safe_sort(
...
     null_pos = np.array([isna(x) for x in values], dtype=bool)
-    nums = np.sort(values[~str_pos & ~null_pos])
-    strs = np.sort(values[str_pos])
-    return np.concatenate(
-        [nums, np.asarray(strs, dtype=object), np.array(values[null_pos])]
-    )
+    num_pos = ~str_pos & ~null_pos
+    str_argsort = np.argsort(values[str_pos])
+    num_argsort = np.argsort(values[num_pos])
+    str_locs = str_pos.nonzero()[0].take(str_argsort)
```
[the remark this answers is not in the document]
Comment (lukemanley) [reply pandas/core/algorithms.py:1903]: it is converting the boolean mask to positional indices within the larger array. we then sort those positional indices via the argsort of the string subset. This is all to be able to call `take` on the original input so we can avoid converting everything to ndarray.

[the remark this answers is not in the document]
Comment (phofl) [reply pandas/core/algorithms.py:1903]: Ah so False=0 and True is nonzero is the catch here? If yes, could you add a short comment?

[the remark this answers is not in the document]
Comment (lukemanley) [reply pandas/core/algorithms.py:1904]: Yes, exactly. I added a comment explaining the operation.

**Review by phofl — APPROVED** [review]
cc @mroeschke

**Comment by mroeschke** [comment]
Just a merge conflict otherwise LGTM

**Pull request merged**